cmd/kubeadm/app/util/config/cluster.go

```diff
@@ -180,23 +183,84 @@ func getNodeNameFromKubeletConfig(kubeconfigDir string) (string, error) {
+		return nil
+	} else {
+		errs = append(errs, errors.WithMessagef(err, "could not retrieve API endpoints for node %q using pod annotations", nodeName))
+	}
+	if err := getAPIEndpointFromClusterStatus(client, nodeName, apiEndpoint); err == nil {
+		return nil
+	} else {
+		errs = append(errs, errors.WithMessagef(err, "could not retrieve API endpoints for node %q using cluster status", nodeName))
+	}
+	return errorsutil.NewAggregate(errs)
+}
+
+func getAPIEndpointFromPodAnnotationWithRetry(client clientset.Interface, nodeName string, apiEndpoint *kubeadmapi.APIEndpoint) error {
+	var rawAPIEndpoint string
+	var lastErr error
+	err := wait.ExponentialBackoff(clientsetretry.DefaultRetry, func() (bool, error) {
+		rawAPIEndpoint, lastErr = getRawAPIEndpointFromPodAnnotation(client, nodeName)
+		return lastErr == nil, nil
+	})
 	if err != nil {
 		return err
 	}
+	parsedAPIEndpoint, err := kubeadmapi.APIEndpointFromString(rawAPIEndpoint)
+	if err != nil {
+		return errors.Wrapf(err, "could not parse API endpoint for node %q", nodeName)
+	}
+	*apiEndpoint = parsedAPIEndpoint
+	return nil
+}
 
-	// gets the APIEndpoint for the current machine from the ClusterStatus
-	e, ok := clusterStatus.APIEndpoints[nodeName]
-	if !ok {
-		return errors.New("failed to get APIEndpoint information for this node")
+func getRawAPIEndpointFromPodAnnotation(client clientset.Interface, nodeName string) (string, error) {
+	podList, err := client.CoreV1().Pods(metav1.NamespaceSystem).List(
+		metav1.ListOptions{
+			FieldSelector: fmt.Sprintf("spec.nodeName=%s", nodeName),
+			LabelSelector: fmt.Sprintf("component=%s,tier=%s", constants.KubeAPIServer, constants.ControlPlaneTier),
+		},
+	)
+	if err != nil {
+		return "", errors.Wrap(err, "could not retrieve list of pods to determine api server endpoints")
+	}
+	if len(podList.Items) == 0 {
+		return "", errors.Errorf("could not find API server pod for node name %q", nodeName)
+	}
+	// Under normal circumstances there's no need to iterate here: we should have only one result.
+	// Upon the strange event of listing more than one apiserver pod for a given node, iterate over
+	// all of them, ignore those missing the annotation, and only error out if we found two different
+	// populated endpoints.
+	rawEndpoint := ""
+	for _, pod := range podList.Items {
+		if podRawEndpoint, ok := pod.Annotations[constants.KubeAPIServerAdvertiseAddressEndpointAnnotationKey]; ok {
+			if len(rawEndpoint) == 0 {
+				rawEndpoint = podRawEndpoint
+			} else if len(podRawEndpoint) > 0 && rawEndpoint != podRawEndpoint {
+				return "", errors.Errorf("found contradictory API server endpoints: %q and %q were found in different pod annotations; cannot continue", rawEndpoint, podRawEndpoint)
+			}
+		}
+		// If there's no annotation at all; ignore this pod
+	}
+	if len(rawEndpoint) == 0 {
+		return "", errors.Errorf("could not find API server endpoint for node name %q", nodeName)
 	}
```

Comment:
@rosti: updated with this strategy, let me know what you think.

I'll add some tests for the utility functions as suggested in the meantime. Thanks!

Reply:
I like your idea, but I was thinking on something simpler. How about just return appropriate errors if the query list does not have a single entry? Then, error out if that single entry does not have annotation.
Let's keep this simple for a start. We can make it more complex if users have problems with it. But for now, it's better to keep it simple as we'll have to support it too.

Reply:
> I like your idea, but I was thinking on something simpler. How about just return appropriate errors if the query list does not have a single entry? Then, error out if that single entry does not have annotation.

I am fine doing this, I was only worried because I haven't experimented enough with pod mirroring and what cases we can expect. Maybe this approach is too strict, but maybe is good enough, I just don't know yet; so yes, sounds fine. I will do this.